### PR TITLE
fix equality of eigen factorizations

### LINKED ISSUE #422: eigen tests fail on nightly
See for example https://github.com/JuliaDiff/ChainRules.jl/pull/421/checks?check_run_id=2697576261:

```
Test Summary:                                            |  Pass  Fail  Broken  Total
ChainRules                                               | 23210     3       4  23217
  rulesets                                               | 23210     3       4  23217
    Base                                                 |  6510             4   6514
    Statistics                                           |    12                   12
    LinearAlgebra                                        | 16643     3          16646
      dense                                              |   585                  585
      norm functions                                     |  1592                 1592
      normalize                                          |    90                   90
      matrix functions                                   |   133                  133
      Structured Matrices                                |   412                  412
      Symmetric/Hermitian rules                          |  3556                 3556
      Factorizations                                     |   505     3            508
        lu decomposition                                 |   159                  159
        svd                                              |   226                  226
        eigendecomposition                               |   104     3            107
          eigen/eigen!                                   |    67     3             70
            eigen!(::Matrix{Float64}) frule              |     7                    7
            eigen!(::Matrix{ComplexF64}) frule           |     4     2              6
              tangents are real when outputs are         |     1                    1
            eigen(::Matrix{Float64}) rrule               |     8                    8
            eigen(::Matrix{ComplexF64}) rrule            |     6     1              7
            normalization/phase functions are idempotent |     4                    4
            normalization/phase functions are idempotent |     4                    4
            hermitian matrices                           |    34                   34
          eigvals/eigvals!                               |    37                   37
        cholesky                                         |    16                   16
      BLAS                                               |  9354                 9354
      LAPACK                                             |   416                  416
    Random                                               |    45                   45
    packages                                             |                      No tests
ERROR: LoadError: Some tests did not pass: 23210 passed, 3 failed, 0 errored, 4 broken.
```

### PR DESCRIPTION
**Pull request by simeonschaub**
At least the newly added field `rcondv` may contain `undef` values, so
these can cause the same eigen factorizations not to compare equal. Not
100% sure whether the other fields should be ignored as well, but since
we didn't have them before, it seems at least consistent to ignore them
here.

fixes JuliaDiff/ChainRules.jl#422

### CODE REVIEW

**Review by vtjnash — COMMENTED** [review]
Does it need to overload `hash` and `isequal` too, currently defined here:
```
julia> methods(hash, (Factorization, UInt))
# 1 method for generic function "hash":
```
https://github.com/JuliaLang/julia/blob/1910a7685a86d44041a98ff874f92e480fc44632/stdlib/LinearAlgebra/src/factorization.jl#L67-L69

**Comment by simeonschaub** [comment]
Oh, good point!

File: stdlib/LinearAlgebra/src/eigen.jl [hunk 1]
```diff
@@ -685,6 +685,16 @@ function show(io::IO, mime::MIME{Symbol("text/plain")}, F::Union{Eigen,Generaliz
     nothing
 end
 
+function Base.hash(F::Eigen, h::UInt)
+    return hash(F.values, hash(F.vectors, h))
```
Comment (simeonschaub) [inline stdlib/LinearAlgebra/src/eigen.jl:689]: Should this hash the type of the factorization as well? I kept it consistent with the previous behavior, but perhaps that should be changed?

Comment (simeonschaub) [reply stdlib/LinearAlgebra/src/eigen.jl]: I have changed this now to include `Eigen` in the hash.

**Comment by simeonschaub** [comment]
Could I get a review for this? Would be nice to fix this, since it currently causes test failures with ChainRules.

**Review by vtjnash — APPROVED** [review]

**Pull request merged**